dbms/src/Processors/Formats/Impl/RegexpRowInputFormat.cpp

```diff
@@ -0,0 +1,155 @@
+    if (format == "Quoted")
+        return FieldFormat::Quoted;
+    if (format == "CSV")
+        return FieldFormat::Csv;
+    if (format == "JSON")
+        return FieldFormat::Json;
+    throw Exception("Unknown field format \"" + format + "\".", ErrorCodes::BAD_ARGUMENTS);
+}
+
+bool RegexpRowInputFormat::readField(size_t index, MutableColumns & columns)
+{
+    const auto & type = getPort().getHeader().getByPosition(index).type;
+    bool parse_as_nullable = format_settings.null_as_default && !type->isNullable();
+    bool read = true;
+    ReadBuffer field_buf(matched_fields[index + 1].first, matched_fields[index + 1].length(), 0);
+    try
+    {
+        switch (field_format)
+        {
+            case FieldFormat::Escaped:
+                if (parse_as_nullable)
+                    read = DataTypeNullable::deserializeTextEscaped(*columns[index], field_buf, format_settings, type);
+                else
+                    type->deserializeAsTextEscaped(*columns[index], field_buf, format_settings);
+                break;
+            case FieldFormat::Quoted:
+                if (parse_as_nullable)
+                    read = DataTypeNullable::deserializeTextQuoted(*columns[index], field_buf, format_settings, type);
+                else
+                    type->deserializeAsTextQuoted(*columns[index], field_buf, format_settings);
+                break;
+            case FieldFormat::Csv:
+                if (parse_as_nullable)
+                    read = DataTypeNullable::deserializeTextCSV(*columns[index], field_buf, format_settings, type);
+                else
+                    type->deserializeAsTextCSV(*columns[index], field_buf, format_settings);
+                break;
+            case FieldFormat::Json:
+                if (parse_as_nullable)
+                    read = DataTypeNullable::deserializeTextJSON(*columns[index], field_buf, format_settings, type);
+                else
+                    type->deserializeAsTextJSON(*columns[index], field_buf, format_settings);
+                break;
+            default:
+                __builtin_unreachable();
+        }
+    }
+    catch (Exception & e)
+    {
+        throw;
+    }
+    return read;
+}
+
+void RegexpRowInputFormat::readFieldsFromMatch(MutableColumns & columns, RowReadExtension & ext)
+{
+    if (matched_fields.size() != columns.size() + 1)
+        throw Exception("The number of matched fields in line doesn't match the number of columns.", ErrorCodes::INCORRECT_DATA);
+
+    ext.read_columns.assign(columns.size(), false);
+    for (size_t columns_index = 0; columns_index < columns.size(); ++columns_index)
+    {
+        ext.read_columns[columns_index] = readField(columns_index, columns);
+    }
+}
+
+bool RegexpRowInputFormat::readRow(MutableColumns & columns, RowReadExtension & ext)
+{
+    if (in.eof())
+        return false;
+
+    char * line_end = find_first_symbols<'\n', '\r'>(in.position(), in.buffer().end());
```

Comment:
If buffer doesn't contain `\n` or `\r`, `line_end` will point to the end of buffer, not the end of line, so it will try to match regexp on some prefix of current line and next call of `readRow(...)` will try to match regexp on some suffix of the same line. You need load entire line in memory (until `\n`, `\r` or EOF). Probably `PeekableReadBuffer` may be useful.